rfcs/0034-expression-integrity.md

```diff
@@ -0,0 +1,142 @@
+---
+feature: expression-integrity
+start-date: 2018-09-28
+author: lrvick
+co-authors:
+related-issues:
+    - https://github.com/NixOS/nix/issues/404
+    - https://github.com/NixOS/nix/issues/613
+    - https://github.com/NixOS/nix/issues/748
+---
+
+# Summary
+[summary]: #summary
+
+This RFC seeks to provide a strategy to allow NixOs to strongly attest who
+authored a nix expression, who reviewed it, and that it has not been tampered
+with outside the review flow.
+
+# Motivation
+[motivation]: #motivation
+
+Due to the lack of VCS integrety on NixOS today a bad actor can gain remote
+code execution on NixOS systems if any of the following are true:
+
+  * A Github employee is coerced or malicious
+  * The Github account credentials of any maintainer are compromised
+  * A successful (even limited) MITM attack on github.com
+
+Essentially NixOS has many single points of trust, and thus single points of
+failure.
+
+This is a serious design flaw and we can learn lessons from other package
+management systems that have been burned by similarly poor package management
+designs.
+
+See examples of major security incidents in other package managers:
+
+ * Gentoo: https://archives.gentoo.org/gentoo-announce/message/dc23d48d2258e1ed91599a8091167002
+ * Debian: https://lists.debian.org/debian-devel-announce/2006/07/msg00003.html
+ * NPM: https://eslint.org/blog/2018/07/postmortem-for-malicious-package-publishes
+ * PyPi: https://www.reddit.com/r/Python/comments/8hvzja/backdoor_in_sshdecorator_package/
+ * Ubuntu Snap: https://github.com/canonical-websites/snapcraft.io/issues/651
+ * Arch Linux AUR: https://lists.archlinux.org/pipermail/aur-general/2018-July/034153.html
+
+# Detailed design
+[design]: #detailed-design
+
+## Package Contributor
+
+### Workflow
+
+  1. Author and test a nix expression
+  2. PR a signed commit adding nix expression to NixOS/nixpkgs repo
+
+### Notes
+
+  * Can be enforced by mandating all commits be signed in VCS settings
+  * Contributors who choose not to sign will need someone else to PR for them
+
+## Package Maintainer
+
+### Workflow
+
+  1. Verify signing key ID is listed in maintainers list
+    * Add key ID to maintainers list if not already present
+  2. Verify signature on PR matches public key id in contributors list
+    * Add key ID to contributors list if not already present
+  3. Review content of new PR for general best practices
+  4. Ensure signatures/hashes verified for third party code referenced
+  5. Make signed merge commit to master of NixOS/nixpkgs
+
+### Notes
+
+  * Maintainer signatures should be a hard requirement
+  * Maintainer and Contributor should never be the same person.
+
+## Nix Clients
+
+### Workflow
+
+  1. Pull latest nix expressions from VCS repo
+  2. Verify author/reviewer commit signatures for all nix expressions
+  3. Build/Install nix expression
+
+### Notes
+
+  * Local building is required for integrity as no trusted cache system exists
```

Comment:
eh?

Reply:
If the current cache system is compromised, it is game over and clients will blindly trust malicious builds. It is only when you have signing by each maintainer -or- a highly transparent reproducible build system that clients will have a way to hold the caching server accountable.

This RFC if implemented would only protect users building from source.

Reply:
I think maybe you misunderstand what our Hydra is? It doesn't host artifacts, it is a build server. Nothing a maintainer builds goes to the binary cache. There is no way for a maintainer to sign the NAR in the end, because the maintainer has no part in its creation. Other distributions do this because they put a great deal of trust in the maintainer to not backdoor the artifacts. We don't, because, well, we don't.

Reply:
My understanding is that it blindly pulls code from github without verifying it, then blindly builds/signs artifacts. This is far worse security than if maintainer/reviewer pairs build for themselves as instead of an individual maintainer/reviewer pairs that can at worst only backdoor their own package, you have a central system that can backdoor any package at any time, and so can those with remote access to that build system.

Because hydra signs anything without code review or verification of signatres from code reviewers it means you need absolute trust in Github, in Hydra, the network that connects them, every person with shell access on the Hydra machines, and every one with push access. This is imo highly irresponsible and honestly puts every individual with that much power in a non obvious amount of personal risk.

As one practical example: If I had a remote shell exec 0-day I would use it gain remote access to hydra to inject a small hard to detect change just before compilation in major bitcoin wallet packages that compromises the random number generator allowing me to factor keys. I would then wait for a big enough wallet balance years later and take it all. I could also blackmail a hydra maintainer to do it for me as the potential payout is worth it. I could also phish the github credentials of one person with commit access, turn off notifications, and rewrite history quickly just after the next change to that package. Hydra can't hope to deliver trusted outputs until it can verify its inputs and independant copies of it can be run with different maintainers that all get the same results (reproducible builds). We can't even begin to solve that problem until we have a trusted signed tree of package definitions without a SPOF.